### PR TITLE
Don't copy into arena when we had to heap allocate for iter

### PR DESCRIPTION
**Pull request by bugadani**
Based on https://github.com/rust-lang/rust/pull/77945#discussion_r506332740

If we already had to heap-allocate for an iterator, let's use that memory instead of copying the contents into the arena. Perf could go either way, but hoping for the best 🤞

cc @est31 since this is your idea (although I did not understand your comment at the time, and I basically came up with the same)

### CODE REVIEW

**Comment by bugadani** [comment]
I think I would prefer some sort of linked list instead of a vector, so that growing that won't re-alloc, but this should be a good enough first approximation of what can be done with this idea. If the tests pass 😅

File: compiler/rustc_arena/src/lib.rs [hunk 1]
```diff
@@ -192,17 +196,24 @@ impl<T> TypedArena<T> {
...
-        }
-        // Move the content to the arena by copying it and then forgetting
-        // the content of the SmallVec
-        unsafe {
-            let len = vec.len();
-            let start_ptr = self.alloc_raw_slice(len);
-            vec.as_ptr().copy_to_nonoverlapping(start_ptr, len);
-            vec.set_len(0);
-            slice::from_raw_parts_mut(start_ptr, len)
+        if vec.len() <= 8 {
```
Comment (est31) [inline compiler/rustc_arena/src/lib.rs:199]: Could probably use a constant here.

Comment (bugadani) [reply compiler/rustc_arena/src/lib.rs:199]: There's `smallvec::inline_size` so I think I'll go with that

File: compiler/rustc_arena/src/lib.rs [hunk 2]
```diff
@@ -44,6 +44,9 @@ pub struct TypedArena<T> {
     /// A vector of arena chunks.
     chunks: RefCell<Vec<TypedArenaChunk<T>>>,
 
+    /// An vector that holds references to heap allocated vectors
```
Comment (Lonami) [inline compiler/rustc_arena/src/lib.rs:47]:
```suggestion
    /// A vector that holds references to heap allocated vectors
```

Comment (bugadani) [reply compiler/rustc_arena/src/lib.rs:47]: Thanks

**Comment by jonas-schievink** [comment]
@rust-timer build c2dea8a743324c28c48d675cb0373cb75d3161f7

**Pull request closed without merging**